### PR TITLE
AEB: Lidar edition

### PR DESCRIPTION
**Pull request by andyblarblar**
This is the draft algorithm for the AEB via LiDAR.

### CODE REVIEW

**Comment by andyblarblar** [comment]
Please use this PR for review on this algo, its definitly not done yet

File: software/AEB.md [hunk 1]
```diff
@@ -8,25 +8,35 @@ is non-deterministic, and thus requires additional deterministic safety systems
...
-        velrec --> vel
-
-    B -- yes --> E[Trigger estop]
-    B -- no --> wait
-```
+- Create an `nxn` grid, where n is the number of voxels, and voxels are some real size measurement `m`.
+- For each LiDAR point in the scan `p`
+  - Get the x,y coordinates of the point relative to the kart by converting from polar to euclidean
+  - If that x,y point is in an unoccupied cell, then mark it as occupied by obstacle (TODO is this trival?)
+- Use the configured wheelbase, steering angle, ect. to create a function for the arc created by the current motion of the kart, herby referred to as `f(t)`
```
Comment (JordanKra) [inline software/AEB.md]: So will this be reading the steering angle of the kart off the can bus? Or will it be inferring steering angle another way?

Comment (andyblarblar) [reply software/AEB.md]: Yes, either through simply assuming the steering commands are correct, or preferably through feedback, if avalible.

File: software/AEB.md [hunk 2]
```diff
@@ -8,25 +8,35 @@ is non-deterministic, and thus requires additional deterministic safety systems
...
+- For each `t`; `0 < t <= some max timestep`
+  - `kart_x, kart_y = f(t)`
+  - Mark that cell containing `kart_x, kart_y` as occupied by kart, call that cell `C`
+  - Mark `c` cells to the left and right of `C` as occupied by the kart, where `c = kart_width/(m*2)`
```
Comment (Bcsotty) [inline software/AEB.md]: For calculation of 'c', we should use the ceiling of the formula stated

File: software/AEB.md [hunk 3]
```diff
@@ -8,25 +8,35 @@ is non-deterministic, and thus requires additional deterministic safety systems
...
+  - If any of those `c` cells attempt to write to a cell that is already occupied by an obstacle (ignoring cells occupied by the kart):
+    - Calculate our time to collision as: `TTC = t` (alternatively, TODO)`TTC = dist(kart_x, kart_y)/kart_vel`, where dist is some distance function from the origin to kart_x, kart_y
+    - If `TTC` > `Tolerance`:
+      - Fire estop, and exit
+- Clear grid, as it will be recalculated with the next lidar scan
 
 ## Functional Requirements
 - REQ1: The ECU will take into account the latest velocity or range estimates available on CAN.
```
Comment (Bcsotty) [inline software/AEB.md:140]: Latest steering angle would also need to be available on CAN.

**Review by Bcsotty — CHANGES_REQUESTED** [review]

File: software/AEB.md [hunk 4]
```diff
@@ -8,30 +8,44 @@ is non-deterministic, and thus requires additional deterministic safety systems
...
-    B -- no --> wait
-```
+- Create an `nxn` grid, where n is the number of voxels, and voxels are some real size measurement `m`.
+- For each LiDAR point in the scan `p`
+  - Get the x,y coordinates of the point relative to the kart by converting from polar to euclidean
+  - If that x,y point is in an unoccupied cell, then mark it as occupied by obstacle (TODO is this trival?)
+- Use the configured wheelbase, steering angle, ect. to create a function for the arc created by the current motion of the kart, herby referred to as `f(t)`
+- For each `t`; `0 < t <= some max timestep`
+  - `kart_x, kart_y = f(t)`
+  - Mark that cell containing `kart_x, kart_y` as occupied by kart, call that cell `C`
```
Comment (Bcsotty) [inline software/AEB.md]: Marking the cell will depend on the size of 'm'. If m is small, then we would need to mark several cells depending on size of kart

Comment (Bcsotty) [reply software/AEB.md]: Another note, the x and y returned by the function will be the x and y position of the center of the rear axle of the kart

Comment (andyblarblar) [reply software/AEB.md]: > Marking the cell will depend on the size of 'm'. If m is small, then we would need to mark several cells depending on size of kart

This is covered by the line below this

Comment (andyblarblar) [reply software/AEB.md]: > Another note, the x and y returned by the function will be the x and y position of the center of the rear axle of the kart

This shouldn't matter I dont think, as we will iteratively fill the voxels with the wirdth of the kart, which will include front to back when done.

**Comment by andyblarblar** [comment]
Current algorithm is fully documented. I'd appreciate you two giving it a read through to make sure its readable and grammitcally sound before we merge.

**Pull request merged**